File: app/src/main/java/com/hgianastasio/biblioulbrav2/views/activities/base/BaseActivity.java
package com.hgianastasio.biblioulbrav2.views.activities.base;

import android.os.Build;
import android.os.Bundle;
import androidx.annotation.CallSuper;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import android.text.Html;
import android.text.method.LinkMovementMethod;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.TextView;

import com.hgianastasio.biblioulbrav2.R;
import com.hgianastasio.biblioulbrav2.di.DaggerApplicationInjector;
import com.hgianastasio.biblioulbrav2.di.components.ActivityComponent;
import com.hgianastasio.biblioulbrav2.di.components.DaggerActivityComponent;
import com.hgianastasio.biblioulbrav2.di.modules.ActivityModule;

import butterknife.BindView;
import butterknife.ButterKnife;

/**
 * Created by heitorgianastasio on 11/29/16.
 */

public abstract class BaseActivity extends AppCompatActivity {

    @BindView(R.id.toolbar)
    Toolbar toolbar;


    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        preBind();
        ButterKnife.bind(this);
        postBind();
    }

    @CallSuper
    protected void preBind(){}

    @CallSuper
    protected void postBind(){
        setSupportActionBar(toolbar);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.default_menu_about,menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId()==R.id.miAbout)
            createAboutDialog().show();
        return super.onOptionsItemSelected(item);
    }

    protected String[] getPermissions(){
        return new String[]{};
    }

    private AlertDialog createAboutDialog(){
        TextView message = (TextView) LayoutInflater.from(this).inflate(R.layout.dialog_about_layout,null);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N)
            message.setText(Html.fromHtml(getString(R.string.about_message,getString(R.string.app_version)),0));
        else
            message.setText(Html.fromHtml(getString(R.string.about_message,getString(R.string.app_version))));
        message.setMovementMethod(LinkMovementMethod.getInstance());
        return new AlertDialog.Builder(this)
                    .setTitle("Sobre")
                    .setView(message)
                    .setNeutralButton("OK",null)
                    .create();
    }

    public final ActivityComponent getActivityComponent(){
        return DaggerActivityComponent
                .builder()
                .activityModule(new ActivityModule(this))
                .applicationComponent(DaggerApplicationInjector.get())
                .build();
    }
}

File: app/src/main/java/com/hgianastasio/biblioulbrav2/views/activities/base/BaseActivity.kt
package com.hgianastasio.biblioulbrav2.views.activities.base

import android.os.Build
import android.os.Bundle
import android.text.Html
import android.text.method.LinkMovementMethod
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuItem
import android.widget.TextView
import androidx.annotation.CallSuper
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import butterknife.BindView
import butterknife.ButterKnife
import com.hgianastasio.biblioulbrav2.R
import com.hgianastasio.biblioulbrav2.di.DaggerApplicationInjector
import com.hgianastasio.biblioulbrav2.di.components.ActivityComponent
import com.hgianastasio.biblioulbrav2.di.components.DaggerActivityComponent
import com.hgianastasio.biblioulbrav2.di.modules.ActivityModule

/**
 * Created by heitorgianastasio on 11/29/16.
 */
abstract class BaseActivity : AppCompatActivity() {
    @kotlin.jvm.JvmField
    @BindView(R.id.toolbar)
    var toolbar: Toolbar? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        preBind()
        ButterKnife.bind(this)
        postBind()
    }

    @CallSuper
    protected open fun preBind() {
    }

    @CallSuper
    protected open fun postBind() {
        setSupportActionBar(toolbar)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.default_menu_about, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.miAbout) createAboutDialog().show()
        return super.onOptionsItemSelected(item)
    }

    protected val permissions: Array<String>
        protected get() = arrayOf()

    private fun createAboutDialog(): AlertDialog {
        val message = LayoutInflater.from(this).inflate(R.layout.dialog_about_layout, null) as TextView
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) message.text = Html.fromHtml(getString(R.string.about_message, getString(R.string.app_version)), 0) else message.text = Html.fromHtml(getString(R.string.about_message, getString(R.string.app_version)))
        message.movementMethod = LinkMovementMethod.getInstance()
        return AlertDialog.Builder(this)
                .setTitle("Sobre")
                .setView(message)
                .setNeutralButton("OK", null)
                .create()
    }

    val activityComponent: ActivityComponent
        get() = DaggerActivityComponent
                .builder()
                .activityModule(ActivityModule(this))
                .applicationComponent(DaggerApplicationInjector.get())
                .build()
}